tools/cli/vz/cmd/login/login.go

```
@@ -0,0 +1,336 @@
+	listener := getFreePort()
+
+	// Generate random code verifier and code challenge pair
+	codeVerifier, codeChallenge := helpers.GenerateRandomCodePair()
+
+	// Generate the redirect uri using the port obtained
+	redirectURI := helpers.GenerateRedirectURI(listener)
+
+	// Generate the login keycloak url by passing the required url parameters
+	loginURL := helpers.GenerateKeycloakAPIURL(codeChallenge,
+		redirectURI,
+	)
+
+	// Busy wait when the authorization code is still not filled by http handle
+	// Close the listener once we obtain it
+	go func() {
+		for authCode == "" {
+
+		}
+		listener.Close()
+	}()
+
+	// Make sure the go routine is running
+	time.Sleep(time.Second)
+
+	// Open the generated keycloak login url in the browser
+	err := helpers.OpenURLInBrowser(loginURL)
+	if err != nil {
+		fmt.Println("Unable to open browser")
+	}
+
+	// Set the handle function and start the http server
+	http.HandleFunc("/",
+		handle,
+	)
+	http.Serve(listener,
+		nil,
+	)
+
+	// Obtain the JWT token by exchanging it with authCode
+	jwtData, err := requestJWT(redirectURI,
+		codeVerifier,
+		caData,
+	)
+	if err != nil {
+		fmt.Println("Unable to obtain the JWT token")
+		return jwtData, err
+	}
+
+	return jwtData, nil
+}
+
+// Obtain the certificate authority data
+// certificate authority data is stored as a secret named system-tls in verrazzano-system namespace
+// Use client cmd to extract the secret
+func extractCAData(kubernetesInterface helpers.Kubernetes) ([]byte, error) {
+	var cert []byte
+
+	kclientset := kubernetesInterface.NewClientSet()
+	secret, err := kclientset.CoreV1().Secrets("verrazzano-system").Get(context.Background(),
+		"system-tls",
+		metav1.GetOptions{},
+	)
+
+	if err != nil {
+		return cert, err
+	}
+	cert = (*secret).Data["ca.crt"]
+	return cert, nil
+}
+
+// Handler function for http server
+// The server page's html,js,etc code is embedded here.
+func handle(w http.ResponseWriter, r *http.Request) {
+	u, _ := url.Parse(r.URL.String())
```

do not suppress errors..

---

okay will change that.